File: Hyperiso/Hyperiso/core/src/Core/adapters/MartyAdapter.cpp
#include "MartyAdapter.h"

#include <stdexcept>

fs::path MartyAdapter::get_path(MartyPath path_name) {
    auto* mm = MemoryManager::GetInstance();

    switch (path_name) {
    case MartyPath::MODEL_FILE: {
        const auto& config = mm->getMemoryCache().config;
        if (!config.mty_model_path.has_value()) {
            throw std::logic_error("MARTY model path is not defined in HyperisoConfig::mty_model_path.");
        }
        return config.mty_model_path.value();
    }

    case MartyPath::PARAM_MAPPING_DIR:
        return mm->get_path(APIPath::PARAM_MAPPING_DIR);

    case MartyPath::SM_MAPPING_FILE:
        return mm->get_path(APIPath::PARAM_MAPPING_DIR) / "sm.json";

    case MartyPath::TEMPLATE_DIR:
        return mm->get_path(APIPath::TEMPLATE_DIR) / "MARTY";

    case MartyPath::MARTY_TEMP_DIR:
        return mm->get_path(APIPath::MARTY_TEMP_DIR);

    case MartyPath::BSM_MAPPING_FILE: {
        auto optional_path = get_optional_path(path_name);
        if (!optional_path.has_value()) {
            throw std::logic_error("MARTY BSM mapping path is not defined in HyperisoConfig::mty_bsm_mapping_path.");
        }
        return optional_path.value();
    }
    }

    throw std::logic_error("Unknown MartyPath.");
}

std::optional<fs::path> MartyAdapter::get_optional_path(MartyPath path_name) {
    if (path_name != MartyPath::BSM_MAPPING_FILE) {
        return get_path(path_name);
    }

    const auto& config = MemoryManager::GetInstance()->getMemoryCache().config;
    return config.mty_bsm_mapping_path;
}

bool MartyAdapter::check_flag(InternalFlag flag) {
    return MemoryManager::GetInstance()->getMemoryCache().flags.at(flag);
}

std::string MartyAdapter::get_marty_model_name() const {
    const auto& config = MemoryManager::GetInstance()->getMemoryCache().config;

    if (!config.mty_model_name.has_value()) {
        throw std::logic_error("MARTY model name is not defined in HyperisoConfig::mty_model_name.");
    }

    return config.mty_model_name.value();
}

File: Hyperiso/Hyperiso/pyhyperiso/core/Core/MartyAdapter.py
"""Python accessors for MARTY-related configuration."""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pyhyperiso.phyperiso.pyhyperiso.core import MartyAdapter as _CppMartyAdapter
from pyhyperiso.phyperiso.pyhyperiso.core import MartyPath as _CppMartyPath


class MartyPath(Enum):
    """MARTY path keys exposed by the C++ runtime.

    Attributes:
        MODEL_FILE: Path to the configured MARTY model file.
        TEMPLATE_DIR: Read-only MARTY template directory.
        PARAM_MAPPING_DIR: Read-only directory containing MARTY/Hyperiso mappings.
        SM_MAPPING_FILE: Read-only SM mapping JSON.
        BSM_MAPPING_FILE: Optional user-provided BSM mapping JSON.
        MARTY_TEMP_DIR: Writable MARTY generated-code/cache directory.
    """

    MODEL_FILE = _CppMartyPath.MODEL_FILE
    TEMPLATE_DIR = _CppMartyPath.TEMPLATE_DIR
    PARAM_MAPPING_DIR = _CppMartyPath.PARAM_MAPPING_DIR
    SM_MAPPING_FILE = _CppMartyPath.SM_MAPPING_FILE
    BSM_MAPPING_FILE = _CppMartyPath.BSM_MAPPING_FILE
    MARTY_TEMP_DIR = _CppMartyPath.MARTY_TEMP_DIR


class MartyAdapter:
    """Inspect MARTY paths and flags used by Hyperiso.

    This adapter is read-only from Python. It is typically useful after
    ``HyperisoMaster.init(...)`` when debugging which MARTY model, mappings,
    templates and writable cache directory were selected.
    """

    def __init__(self) -> None:
        """Create an adapter bound to the current C++ MARTY configuration."""
        self._cpp_obj = _CppMartyAdapter()

    def get_path(self, path_kind: MartyPath) -> str:
        """Return a required MARTY path by kind."""
        return str(self._cpp_obj.get_path(path_kind.value))

    def get_optional_path(self, path_kind: MartyPath) -> Optional[str]:
        """Return an optional MARTY path, or ``None`` when it is not configured."""
        value = self._cpp_obj.get_optional_path(path_kind.value)
        return None if value is None else str(value)

    def check_flag(self, flag) -> bool:
        """Return the value of a MARTY-related C++ flag."""
        return bool(self._cpp_obj.check_flag(flag))

    def get_model_name(self) -> str:
        """Return the configured MARTY model name."""
        return str(self._cpp_obj.get_marty_model_name())

    def __repr__(self) -> str:
        """Return a compact representation including the model name."""
        return f"<PyMartyAdapter model='{self.get_model_name()}'>"


__all__ = ["MartyAdapter", "MartyPath"]
